gcloud/storage/api.py

```diff
@@ -184,6 +170,11 @@ class _BucketIterator(Iterator):
     """
 
     def __init__(self, connection, extra_params=None):
+        # If an implicit connection was intended, we pass along `None` to the
+        # Bucket() constructor as well.
```

Comment:
Yes. This part is the goofiest. I think after writing #761 the approach of #728 is more feasible. As you mentioned, we have many methods that use a connection (i.e. make HTTP requests) that don't need to.

The hoop-jumping here is certainly going to confuse users.